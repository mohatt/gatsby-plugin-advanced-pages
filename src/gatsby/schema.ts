import type { PluginOptionsSchemaJoi } from 'gatsby-plugin-utils'

/**
 * Defines the schema for page definitions.
 *
 * @internal
 */
export const getPagesSchema = (Joi: PluginOptionsSchemaJoi) => {
  return Joi.array()
    .items(
      Joi.object({
        title: Joi.string().required(),
        template: Joi.string(),
        helper: Joi.string(),
        routes: Joi.object({})
          .required()
          .min(1)
          .unknown(true)
          .pattern(
            Joi.any(),
            Joi.string()
              .pattern(/^\//)
              .message("{{#label}}: Route path must be a string starting with '/'"),
          ),
        data: Joi.object({}).default({}).unknown(true),
      }),
    )
    .label('root')
    .default([])
}

/**
 * Defines the schema for plugin options.
 *
 * @internal
 */
export const getOptionsSchema = (Joi: PluginOptionsSchemaJoi) => {
  return Joi.object({
    basePath: Joi.string().description('Root url for all pages created through the plugin.'),
    pages: getPagesSchema(Joi).description(
      'Inline pages configuration to use instead of pages.config.js.',
    ),
    template: Joi.string().description(
      'Default template to be used for pages with no "template" metadata defined.',
    ),
    directories: Joi.object({
      templates: Joi.string().description('Location of template components used to render pages.'),
      helpers: Joi.string().description('Location of page helpers.'),
    }),
    pagination: Joi.object({
      limit: Joi.number()
        .positive()
        .description(
          'Default page size to be used when no "limit" parameter is passed to "createAdvancedPage()".',
        ),
      suffix: Joi.string()
        .pattern(/\/:page(\/|$)/)
        .description(
          'Suffix to be added to the original route to generate a paginated route. This is only used when no paginated route is passed to "createAdvancedPage()".',
        )
        .messages({
          'string.pattern.base':
            "{{#label}} with value \"{{#value}}\": Route suffix must contain the ':page' paramater (e.g '/pages/:page')",
        }),
    }),
    typeNames: Joi.object({
      page: Joi.string().description('Name of the Page object type.'),
      pageRoute: Joi.string().description('Name of the Page Route object type.'),
    }),
  })
}
