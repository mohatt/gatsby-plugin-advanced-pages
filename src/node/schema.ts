import type { PluginOptionsSchemaJoi } from 'gatsby-plugin-utils'

/**
 * Defines the schema for page definitions.
 *
 * @internal
 */
export const getPagesSchema = (Joi: PluginOptionsSchemaJoi) => {
  return Joi.array().items(
    Joi.object({
      title: Joi.string().required(),
      template: Joi.string(),
      helper: Joi.string(),
      routes: Joi.object({})
        .required()
        .min(1)
        .unknown(true)
        .pattern(
          Joi.any(),
          Joi.string()
            .pattern(/^\//)
            .message("{{#label}}: Route paths must be strings that start with '/'"),
        ),
      data: Joi.object({}).default({}).unknown(true),
    }),
  )
}

/**
 * Defines the schema for plugin options.
 *
 * @internal
 */
export const getOptionsSchema = (Joi: PluginOptionsSchemaJoi) => {
  return Joi.object({
    basePath: Joi.string().description('Root URL for all pages created using this plugin.'),
    pages: getPagesSchema(Joi)
      .default([])
      .description('Inline pages configuration, used instead of a separate pages.config.* file.'),
    template: Joi.string().description(
      'Default template to be used when no "template" metadata is provided.',
    ),
    directories: Joi.object({
      templates: Joi.string().description('Path to template components used to render pages.'),
      helpers: Joi.string().description('Path to page helper functions.'),
    }),
    pagination: Joi.object({
      limit: Joi.number()
        .positive()
        .description(
          'Default number of items per page when "limit" is not explicitly provided in "createAdvancedPage()".',
        ),
      suffix: Joi.string()
        .pattern(/\/:page(\/|$)/)
        .description(
          'Suffix added to a route to generate paginated paths. Used only when no custom pagination route is provided in "createAdvancedPage()".',
        )
        .messages({
          'string.pattern.base':
            "{{#label}} with value \"{{#value}}\" must contain ':page' (e.g., '/pages/:page') to support pagination.",
        }),
    }),
    typeNames: Joi.object({
      page: Joi.string().description('GraphQL type name for pages.'),
      pageRoute: Joi.string().description('GraphQL type name for page routes.'),
    }),
  })
}
