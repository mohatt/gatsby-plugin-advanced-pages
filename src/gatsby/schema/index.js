export { default as optionsSchema } from './options'
export { default as pagesSchema } from './pages'
