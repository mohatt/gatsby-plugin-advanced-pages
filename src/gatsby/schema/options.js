export default function (Joi) {
  return Joi.object({
    basePath: Joi.string()
      .default('/')
      .description(`Root url for all pages created through the plugin.`),
    template: Joi.string()
      .default(null)
      .description(`Default template to be used for pages with no "template" metadata defined.`),
    directories: Joi.object({
      pages: Joi.string()
        .default('.')
        .description(`Location to look for the pages config file.`),
      templates: Joi.string()
        .default('./src/templates')
        .description(`Location of template components used to render pages.`),
      helpers: Joi.string()
        .default('./gatsby/pages')
        .description(`Location of page helpers.`),
    }).default(),
    pagination: Joi.object({
      limit: Joi.number()
        .positive()
        .default(10)
        .description(`Default page size to be used when no "limit" parameter is passed to "createAdvancedPage()".`),
      suffix: Joi.string()
        .pattern(/\/\:page(\/|$)/)
        .default('/page/:page')
        .description(`Suffix to be added to the original route to generate a paginated route. This is only used when no paginated route is passed to "createAdvancedPage()".`)
        .messages({
          'string.pattern.base': `{{#label}} with value "{{#value}}": Route suffix must contain the ':page' paramater (e.g '/pages/:page')`
        }),
    }).default(),
    typeNames: Joi.object({
      page: Joi.string()
        .pattern(/^[A-Z]\w+$/)
        .default('Page')
        .description(`Name of the Page object type.`)
        .messages({
          'string.pattern.base': '{{#label}} with value "{{#value}}": Object type must be Title-cased'
        }),
    }).default(),
  })
}
