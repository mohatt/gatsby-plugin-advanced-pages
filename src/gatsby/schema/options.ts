import type { PluginOptionsSchemaJoi } from 'gatsby-plugin-utils'
import getPagesSchema from './pages'

const getOptionsSchema = (Joi: PluginOptionsSchemaJoi) => {
  return Joi.object({
    basePath: Joi.string()
      .description('Root url for all pages created through the plugin.'),
    pages: getPagesSchema(Joi)
      .description('Inline pages configuration to use instead of pages.config.js.'),
    template: Joi.string()
      .description('Default template to be used for pages with no "template" metadata defined.'),
    directories: Joi.object({
      templates: Joi.string()
        .description('Location of template components used to render pages.'),
      helpers: Joi.string()
        .description('Location of page helpers.')
    }),
    pagination: Joi.object({
      limit: Joi.number()
        .positive()
        .description('Default page size to be used when no "limit" parameter is passed to "createAdvancedPage()".'),
      suffix: Joi.string()
        .pattern(/\/:page(\/|$)/)
        .description('Suffix to be added to the original route to generate a paginated route. This is only used when no paginated route is passed to "createAdvancedPage()".')
        .messages({
          'string.pattern.base': '{{#label}} with value "{{#value}}": Route suffix must contain the \':page\' paramater (e.g \'/pages/:page\')'
        })
    }),
    typeNames: Joi.object({
      page: Joi.string()
        .description('Name of the Page object type.')
    })
  })
}

export default getOptionsSchema
