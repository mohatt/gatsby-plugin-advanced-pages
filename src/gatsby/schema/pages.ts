import type { PluginOptionsSchemaJoi } from 'gatsby-plugin-utils'

const getPagesSchema = (Joi: PluginOptionsSchemaJoi) => {
  return Joi.array().items(
    Joi.object({
      title: Joi.string().required(),
      template: Joi.string(),
      helper: Joi.string(),
      routes: Joi.object({})
        .required()
        .min(1)
        .unknown(true)
        .pattern(Joi.any(), Joi.string()
          .pattern(/^\//)
          .message('{{#label}}: Route path must be a string starting with \'/\'')
        ),
      data: Joi.object({})
        .default({})
        .unknown(true)
    })
  ).label('root').default([])
}

export default getPagesSchema
