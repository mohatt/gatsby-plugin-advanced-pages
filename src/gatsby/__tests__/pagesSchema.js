import { testPluginOptionsSchema } from 'gatsby-plugin-utils'
import { pagesSchema } from '../schema'
import testCases from '../../../test/__fixtures__/pagesSchema'

describe('pluginOptionsSchema', () => {
  for (const { title, pages } of testCases) {
    it(title, async () => {
      const { isValid, errors } = await testPluginOptionsSchema(
        ({ Joi }) => pagesSchema(Joi), pages
      )

      expect(isValid).toMatchSnapshot()
      expect(errors).toMatchSnapshot()
    })
  }
})
