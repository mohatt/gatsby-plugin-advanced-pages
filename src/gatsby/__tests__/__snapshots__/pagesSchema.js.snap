// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pluginOptionsSchema should invalidate empty fields 1`] = `false`;

exports[`pluginOptionsSchema should invalidate empty fields 2`] = `
Array [
  "\\"[0].title\\" is not allowed to be empty",
  "\\"[0].template\\" is not allowed to be empty",
  "\\"[0].helper\\" is not allowed to be empty",
  "\\"[0].routes\\" must have at least 1 key",
]
`;

exports[`pluginOptionsSchema should invalidate fields with incorrect data types 1`] = `false`;

exports[`pluginOptionsSchema should invalidate fields with incorrect data types 2`] = `
Array [
  "\\"[0].title\\" must be a string",
  "\\"[0].template\\" must be a string",
  "\\"[0].helper\\" must be a string",
  "\\"[0].routes\\" must be of type object",
  "\\"[0].data\\" must be of type object",
]
`;

exports[`pluginOptionsSchema should invalidate required fields with missing values 1`] = `false`;

exports[`pluginOptionsSchema should invalidate required fields with missing values 2`] = `
Array [
  "\\"[0].title\\" is required",
  "\\"[0].routes\\" is required",
]
`;

exports[`pluginOptionsSchema should invalidate routes with invalid values 1`] = `false`;

exports[`pluginOptionsSchema should invalidate routes with invalid values 2`] = `
Array [
  "\\"[0].routes.foo\\": Route path must be a string starting with '/'",
  "\\"[1].routes.bar\\" is not allowed to be empty",
  "\\"[2].routes.third\\" must be a string",
]
`;

exports[`pluginOptionsSchema should invalidate unknown fields  1`] = `false`;

exports[`pluginOptionsSchema should invalidate unknown fields  2`] = `
Array [
  "\\"[0].foo\\" is not allowed",
]
`;

exports[`pluginOptionsSchema should validate correct fields 1`] = `true`;

exports[`pluginOptionsSchema should validate correct fields 2`] = `Array []`;
